providers/implementations/ciphers/cipher_aes_xts.c

```diff
@@ -137,6 +149,11 @@ static void *aes_xts_dupctx(void *vctx)
     if (!ossl_prov_is_running())
         return NULL;
 
+#ifdef AES_XTS_S390X
+    if (in->plat.s390x.fc)
+        return s390x_aes_xts_dupctx(vctx);
```

Comment:
What if one duplicates the context after setting the IV, but before setting the key? In this case `in->plat.s390x.fc` is not yet set (it will only be set when setting the key), and thus you would not duplicate the platform specific things, especially, you would miss to copy the IV that was set in `km->tweak`, and thus the duplicated context will not have `km->tweak` set as well as `xctx->plat.s390x.iv_set` would remain zero.

Why not always copying the entire `plat` structure when AES_XTS_S390X is defined, regardless if `fc` is set or not?

Reply:
You're right. This needs to be fixed before merging the stuff. 

Another option could be to setup the `fc` based on `xctx->base.keylen`, even if `key` is `NULL`. I would prefer this solution over your proposal, because if will skip some `memcpy()` in case the fc is not supported by the hardware.

Reply:
I debugged this case (init iv with type and iv, dup and set key to both ctx) and it looks like it's working correctly. A closer look into the code shows, that `s390x_aes_xts_dup()` is skipped (as expected), but `cipher_hw_aes_xts_copyctx()` is called instead and this function is also doing a full copy of the context. So its working, but only by chance.

As mentioned, I'm working on a solution by setting the fc based on the `xctx->base.keylen`. This is in my opinion the right way to do, as it does not rely on other code paths are doing it right.